### PR TITLE
Pressure Dependence Improvements

### PR DESCRIPTION
**Pull request by mjohnson541**
This PR adds the simulation least squares methods and the Georgievskii et al. 2013 CSE variant to Arkane and makes a number of improvements to pdep sensitivity analysis and the explorer tool.

### CODE REVIEW

File: rmgpy/pdep/me.pyx [hunk 1]
```diff
@@ -39,21 +39,21 @@ import rmgpy.constants as constants
 ################################################################################
 
 
-cpdef generate_full_me_matrix(network, bint products=True):
+cpdef generate_full_me_matrix(network, bint products=True, bint exclude_association=False, bint neglect_high_energy_collisions=False, high_energy_rate_tol=0.01):
```
Comment (hwpang) [inline rmgpy/pdep/me.pyx:42]: Does high_energy_rate_tol not need a typing in front of it?

File: rmgpy/pdep/me.pyx [hunk 2]
```diff
@@ -123,10 +123,42 @@ cpdef generate_full_me_matrix(network, bint products=True):
...
+                                    me_mat[ind,ind] -= me_mat[ind1,ind]
+                                    me_mat[ind1,ind] = 0.0
+                                    me_mat[ind,ind1] = 0.0
+
+                            me_mat[ind1,ind1] = -ktot
+
+    if exclude_association:
+        return me_mat[:-(n_reac+n_prod),:-(n_reac+n_prod)], indices
+    else:
     return me_mat, indices
```
Comment (hwpang) [inline rmgpy/pdep/me.pyx]: Indentation?

File: rmgpy/pdep/cse.pyx [hunk 3]
```diff
@@ -223,3 +223,163 @@ def apply_chemically_significant_eigenvalues_method(network, list lumping_order=
 
     # Return the matrix of k(T,P) values and the pseudo-steady population distributions
     return k, pa
+
+def get_rate_coefficients_CSE_Advanced(network,T,P,neglect_high_energy_collisions=False,high_energy_rate_tol=0.01):
```
Comment (hwpang) [inline rmgpy/pdep/cse.pyx:228]: blank after comma

File: rmgpy/pdep/cse.pyx [hunk 4]
```diff
@@ -223,3 +223,163 @@ def apply_chemically_significant_eigenvalues_method(network, list lumping_order=
...
+    """
+    CSE using the Georgievskii et al. 2013 method
```
Comment (hwpang) [inline rmgpy/pdep/cse.pyx:230]: Maybe put the full reference here for future developers

File: rmgpy/pdep/cse.pyx [hunk 5]
```diff
@@ -223,3 +223,163 @@ def apply_chemically_significant_eigenvalues_method(network, list lumping_order=
...
+                kmat[j,i] = 0.0
+                kmat[j,i] = (W @ pvlmbd[:n_cse,i-n_isom])[j]
+
+            elif len(prod.species) == 2:
+                kmat[j,i] = 0.0
+                kmat[j,i] = (Gvlmbd[:n_cse,j-n_isom] @ Winv[:,i])
+
+    return kmat
+
+def apply_chemically_significant_eigenvalues_method_georgievskii(network, neglect_high_energy_collisions=neglect_high_energy_collisions,
```
Comment (hwpang) [inline rmgpy/pdep/cse.pyx]: Default values undefined here?

File: rmgpy/pdep/cse.pyx [hunk 6]
```diff
@@ -223,3 +223,163 @@ def apply_chemically_significant_eigenvalues_method(network, list lumping_order=
...
+                                                                 high_energy_rate_tol=high_energy_rate_tol):
+    return get_rate_coefficients_CSE_Advanced(network, network.T, network.P, neglect_high_energy_collisions=False,
```
Comment (hwpang) [inline rmgpy/pdep/cse.pyx]: Should use the variables instead of hard-coded?

File: rmgpy/pdep/sls.py [hunk 7]
```diff
@@ -42,3 +42,36 @@
 from rmgpy.pdep.reaction import calculate_microcanonical_rate_coefficient
 from rmgpy.pdep.me import generate_full_me_matrix
 from rmgpy.statmech.translation import IdealGasTranslation
+
+def get_initial_condition(network,x0,indices):
```
Comment (hwpang) [inline rmgpy/pdep/sls.py:47]: Could you add blanks after commas

File: rmgpy/pdep/sls.py [hunk 8]
```diff
@@ -75,3 +75,25 @@ def get_initial_condition(network,x0,indices):
...
+    return M*u
+end""")
+    jac = Main.eval("""
+function jac(u,M,t)
+    return M
+end""")
+    tspan = (0.0,t)
+    fcn = de.ODEFunction(f,jac=jac)
+    prob = de.ODEProblem(fcn,p0,tspan,M)
+    sol = de.solve(prob,solver=de.CVODE_BDF(),abstol=1e-16,reltol=1e-6)
```
Comment (hwpang) [inline rmgpy/pdep/sls.py:95]: Would it be useful to allow abstol and reltol as unser inputs?

File: rmgpy/pdep/sls.py [hunk 9]
```diff
@@ -101,3 +101,25 @@ def solve_me_fcns(f,jac,M,p0,t):
     sol = de.solve(prob,solver=de.CVODE_BDF(),abstol=1e-16,reltol=1e-6)
     return sol
 
+def states_to_configurations(network,indices,state,exclude_association=False):
```
Comment (hwpang) [inline rmgpy/pdep/sls.py]: Could you add blanks after commas

File: rmgpy/pdep/sls.py [hunk 10]
```diff
@@ -123,3 +123,53 @@ def states_to_configurations(network,indices,state,exclude_association=False):
             xs[i+network.n_isom] += state[-network.n_reac-network.n_prod+i]
     return xs
 
+def ravel_kmat(kmat,n_isomreac):
```
Comment (hwpang) [inline rmgpy/pdep/sls.py:320]: Could you add blanks after commas for all functions here?

File: rmgpy/pdep/sls.py [hunk 11]
```diff
@@ -123,6 +123,208 @@ def states_to_configurations(network,indices,state,exclude_association=False):
             xs[i+network.n_isom] += state[-network.n_reac-network.n_prod+i]
     return xs
 
+def get_rate_coefficients_SLS(network,T,P,method="mexp",neglect_high_energy_collisions=False,high_energy_rate_tol=0.01):
+    """
+    Generate phenomenological rate coefficients using the simulation least squares method
+    """
```
Comment (hwpang) [inline rmgpy/pdep/sls.py:108]: Maybe add all available methods in the docstring

File: rmgpy/pdep/cse.pyx [hunk 12]
```diff
@@ -343,7 +343,7 @@ def get_rate_coefficients_CSE_Advanced(network,T,P,neglect_high_energy_collision
 
 
     pvlmbd = np.linalg.solve(eigen_vectors,Pv)
-    Gvlmbd = eigen_vectors.T @ k2v
```
Comment (hwpang) [inline rmgpy/pdep/cse.pyx]: Could you help me understand why is this necessary?

File: arkane/pdep.py [hunk 13]
```diff
@@ -107,6 +107,17 @@ def __init__(self, network,
                  maximumGrainSize=None, minimumGrainCount=0,
                  method=None, interpolationModel=None, maximumAtoms=None,
                  activeKRotor=True, activeJRotor=True, rmgmode=False, sensitivity_conditions=None):
+
+        if len(network.products) > 0:
+            if method in "simulation least squares":
```
Comment (hwpang) [inline arkane/pdep.py:113]: Why are we using in here?

File: arkane/pdep.py [hunk 14]
```diff
@@ -297,7 +297,7 @@ def execute(self, output_file, plot, file_format='pdf', print_summary=True):
                 self.plot(os.path.dirname(output_file))
             if self.sensitivity_conditions is not None:
                 logging.info('\n\nRunning sensitivity analysis...')
-                SensAnalysis(self, os.path.dirname(output_file), perturbation=self.sensitivity_perturbation)
+                SensAnalysis(deepcopy(self), os.path.dirname(output_file), perturbation=self.sensitivity_perturbation)
```
Comment (hwpang) [inline arkane/pdep.py:300]: Should this be assigned to anything? Or does it change things in place?

**Review by hwpang — CHANGES_REQUESTED** [review]
Thanks for the PR! I have a few comments. I'm not done reading all the commits yet, and I will come back another time!

**Comment by hwpang** [comment]
@mjohnson541 Also could you rebase?

File: rmgpy/pdep/sls.py (hunk 8 above)
Comment (mjohnson541) [reply rmgpy/pdep/sls.py:95]: We could, however, adding a lever like that implies that users should be adjusting it. Right now I think our recommendation is to use the matrix exponential algorithm and the ODE method really only exists (beyond I put it in the paper so it should probably be made available) for users to prove to themselves that the matrix exponential method is working XD. I think the given tolerances should be sufficient for that purpose and if someone is running into sensitive numerics they're going to need to dig around the code anyway.

File: rmgpy/pdep/cse.pyx (hunk 12 above)
Comment (mjohnson541) [reply rmgpy/pdep/cse.pyx]: I believe this is identical. So strictly speaking unnecessary XD.

File: arkane/pdep.py (hunk 13 above)
Comment (mjohnson541) [reply arkane/pdep.py:113]: Sorry, should be flipped. We're looking to see if "simulation least squares" is within the method name.

File: arkane/pdep.py (hunk 14 above)
Comment (mjohnson541) [reply arkane/pdep.py:300]: It doesn't change anything in place, but it executes on initialization after the associated main pdep job is finished so it doesn't matter that it isn't assigned to anything.

File: rmgpy/pdep/me.pyx (hunk 1 above)
Comment (mjohnson541) [reply rmgpy/pdep/me.pyx]: fixed

File: rmgpy/pdep/me.pyx (hunk 2 above)
Comment (mjohnson541) [reply rmgpy/pdep/me.pyx]: fixed

File: rmgpy/pdep/cse.pyx (hunk 3 above)
Comment (mjohnson541) [reply rmgpy/pdep/cse.pyx]: fixed

File: rmgpy/pdep/cse.pyx (hunk 4 above)
Comment (mjohnson541) [reply rmgpy/pdep/cse.pyx]: added

File: rmgpy/pdep/cse.pyx (hunk 5 above)
Comment (mjohnson541) [reply rmgpy/pdep/cse.pyx]: added

File: rmgpy/pdep/cse.pyx (hunk 6 above)
Comment (mjohnson541) [reply rmgpy/pdep/cse.pyx]: fixed

File: rmgpy/pdep/sls.py (hunk 7 above)
Comment (mjohnson541) [reply rmgpy/pdep/sls.py]: added

File: rmgpy/pdep/sls.py (hunk 9 above)
Comment (mjohnson541) [reply rmgpy/pdep/sls.py]: added

File: rmgpy/pdep/sls.py (hunk 11 above)
Comment (mjohnson541) [reply rmgpy/pdep/sls.py:111]: added

File: rmgpy/pdep/cse.pyx (hunk 12 above)
Comment (mjohnson541) [reply rmgpy/pdep/cse.pyx]: I think I got rid of this.

**Comment by mjohnson541** [comment]
Thanks @hwpang I believe I have fixed all the issues now.

File: rmgpy/pdep/me.pyx (hunk 2 above)
Comment (hwpang) [reply rmgpy/pdep/me.pyx]: I think the indentation here is still not right

File: rmgpy/pdep/cse.pyx [hunk 15]
```diff
@@ -217,9 +217,168 @@ def apply_chemically_significant_eigenvalues_method(network, list lumping_order=
...
+                                            * exp(-e_list[r] / (constants.R * temperature)) * eq_ratios[i])
+                    s_mat_inv[index] = 1.0 / s_mat[index]
+
+    # Symmetrize master equation matrix: me_mat = s_mat * m_symm * s_mat_inv
+    # Since s_mat and s_mat_inv are diagonal we can do this very efficiently
+    for r in range(n_rows):
+        for s in range(n_rows):
+            me_mat[r, s] = s_mat_inv[r] * me_mat[r, s] * s_mat[s]
+
+    # DEBUG: Check that the matrix has been properly symmetrized
```
Comment (hwpang) [inline rmgpy/pdep/cse.pyx:284]: Why DEBUG here?

File: rmgpy/pdep/me.pyx (hunk 2 above)
Comment (hwpang) [reply rmgpy/pdep/me.pyx]: If you fix this in another commit, could you squash it?

File: rmgpy/pdep/sls.py (hunk 9 above)
Comment (hwpang) [reply rmgpy/pdep/sls.py]: Could you squash the commit where you add the blanks? Thanks

File: rmgpy/pdep/sls.py [hunk 16]
```diff
@@ -101,27 +101,210 @@ def solve_me_fcns(f, jac, M, p0, t):
...
-    sum full master equation state into total species concentrations
+    Generate phenomenological rate coefficients using the simulation least squares method
+    valid methods: "mexp" (matrix exponential (default)),
+                   "ode" (ordinary differential equation solve)
+                   "eigen" (eigendecomposition)
     """
-    if exclude_association:
-        xs = np.zeros(network.n_isom)
+    if network.T != T or network.P != P:
+        network.set_conditions(T,P)
```
Comment (hwpang) [inline rmgpy/pdep/sls.py:113]: Could you add blank after comma here

File: rmgpy/pdep/sls.py [hunk 17]
```diff
@@ -101,27 +101,210 @@ def solve_me_fcns(f, jac, M, p0, t):
...
-                index = indices[i,r,s]
-                if index == -1:
-                    continue
-                cum += state[index]
-        xs[i] += cum
-    if not exclude_association:
-        for i in range(network.n_reac+network.n_prod):
-            xs[i+network.n_isom] += state[-network.n_reac-network.n_prod+i]
-    return xs
+        raise ValueError
```
Comment (hwpang) [inline rmgpy/pdep/sls.py:220]: Maybe give an error message here

File: rmgpy/pdep/sls.py [hunk 18]
```diff
@@ -101,27 +101,210 @@ def solve_me_fcns(f, jac, M, p0, t):
...
+    products = network.products[:]
+    xsorder = isomers+reactants+products
+    n_isomreac = len(isomers+reactants)
+    xssource = []
+    dxdtssource = []
+    xseq = []
+    dxdtseq = []
+
+    #Single domainant source simulations
+    for i,isomer in enumerate(isomers):
```
Comment (hwpang) [inline rmgpy/pdep/sls.py:233]: Could you add blank after commas in this commit? Also the other ones in this commit

File: rmgpy/pdep/cse.pyx (hunk 12 above)
Comment (hwpang) [reply rmgpy/pdep/cse.pyx]: Oh, could you squash that one commit that fixes this with this one?

File: arkane/sensitivity.py [hunk 19]
```diff
@@ -454,7 +454,10 @@ def plot(self, wells, transition_states):
                 axis.set_xlabel(r'Sensitivity: $\frac{\partial\:\ln{k}}{\partial\:E0}$, ($\frac{J}{mol}$)')
                 # axis.ticklabel_format('sci')
                 axis.set_title('{0}, {1}'.format(condition[0], condition[1]))
-                axis.set_xlim([min_sa, max_sa])
+                try:
+                    axis.set_xlim([min_sa, max_sa])
+                except:
+                    pass
```
Comment (hwpang) [inline arkane/sensitivity.py:484]: Should we be throwing warning with NaN sensitivity values or are we throwing it somewhere else already?

File: arkane/pdep.py (hunk 14 above)
Comment (hwpang) [reply arkane/pdep.py:300]: I'm confused, could you help me understand why a copy of self is needed here if it doesn't change things in place?

File: arkane/sensitivity.py [hunk 20]
```diff
@@ -266,6 +266,30 @@ def __init__(self, job, output_directory, perturbation, max_iters=5):
...
+        base_wells = []
+        base_wells.extend(self.job.network.reactants)
+        base_wells.extend(self.job.network.isomers)
+        base_wells.extend(self.job.network.products)
+        Emin = min([x.E0 for x in base_wells])
+        
+        base_transition_states = []
+        Emax = Emin
+        for rxn in self.job.network.path_reactions:
+            # if rxn.transition_state is not None:
```
Comment (hwpang) [inline arkane/sensitivity.py:280]: Why is this commented?

File: rmgpy/pdep/me.pyx (hunk 2 above)
Comment (mjohnson541) [reply rmgpy/pdep/me.pyx]: fixed

File: rmgpy/pdep/sls.py (hunk 9 above)
Comment (mjohnson541) [reply rmgpy/pdep/sls.py]: This shouldn't exist in the first place anymore

File: rmgpy/pdep/cse.pyx (hunk 12 above)
Comment (mjohnson541) [reply rmgpy/pdep/cse.pyx]: fixed

File: arkane/pdep.py (hunk 14 above)
Comment (mjohnson541) [reply arkane/pdep.py:300]: So I made this commit years ago so I don't precisely remember, but SensAnalysis relies on modifying the network to do what it does so my guess is that the issue is that if you don't copy it it saves itself incorrectly.

File: rmgpy/pdep/cse.pyx (hunk 15 above)
Comment (mjohnson541) [reply rmgpy/pdep/cse.pyx:284]: This code segment is copied from the vanilla cse method. Fundamentally, it's running a code check that should be true there so I guess they called it a debugging opeeration.

File: arkane/sensitivity.py (hunk 19 above)
Comment (mjohnson541) [reply arkane/sensitivity.py:484]: Plotting happens last so I wouldn't worry too much about NaN handling. If plotting fails they'll be able to see the NaN's in the sensitivity output.

File: arkane/sensitivity.py (hunk 20 above)
Comment (mjohnson541) [reply arkane/sensitivity.py:280]: removed

File: rmgpy/pdep/sls.py (hunk 18 above)
Comment (mjohnson541) [reply rmgpy/pdep/sls.py:233]: done

File: rmgpy/pdep/sls.py (hunk 17 above)
Comment (mjohnson541) [reply rmgpy/pdep/sls.py:220]: added

File: rmgpy/pdep/sls.py [hunk 21]
```diff
@@ -79,3 +79,25 @@ def get_initial_condition(network, x0, indices):
...
+    f = Main.eval("""
+function f(u,M,t)
+    return M*u
+end""")
+    jac = Main.eval("""
+function jac(u,M,t)
+    return M
+end""")
+    tspan = (0.0,t)
+    fcn = de.ODEFunction(f,jac=jac)
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/pdep/sls.py:93]: added

File: rmgpy/pdep/network.py [hunk 22]
```diff
@@ -267,10 +268,28 @@ def calculate_rate_coefficients(self, Tlist, Plist, method, error_check=True):
                 elif method.lower() == 'reservoir state':
                     self.apply_reservoir_state_method()
                 elif method.lower() == 'chemically-significant eigenvalues':
-                    self.apply_chemically_significant_eigenvalues_method()
+                    self.apply_chemically_significant_eigenvalues_method(method="allen",neglect_high_energy_collisions=neglect_high_energy_collisions,
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply rmgpy/pdep/network.py:276]: added

**Review by hwpang — APPROVED** [review]
@mjohnson541 Thanks for the PR and changes! Looks good to me!
